File: src/vpin.ts
/**
 * VPIN — Volume-Synchronized Probability of Informed Trading.
 *
 * Easley, López de Prado & O'Hara (2012), "Flow Toxicity and Liquidity in a
 * High-Frequency World". Trades are grouped into equal-volume buckets; each
 * bucket is split into buy/sell volume by Bulk Volume Classification (BVC)
 * using the standardized close-to-close price change; VPIN is the average
 * absolute order imbalance across a rolling window of buckets.
 *
 *   V_buy  = V · Φ(ΔP / σ)
 *   V_sell = V · (1 − Φ(ΔP / σ))
 *   VPIN   = Σ |V_buy − V_sell| / Σ V     over the window
 */

export interface VolumeBucket {
  /** close-to-close price change over the bucket */
  priceChange: number;
  /** total traded volume in the bucket */
  volume: number;
}

// Abramowitz & Stegun 7.1.26 — max error ≈ 1.5e-7.
function erf(x: number): number {
  const t = 1 / (1 + 0.3275911 * Math.abs(x));
  const y =
    1 -
    ((((1.061405429 * t - 1.453152027) * t + 1.421413741) * t - 0.284496736) *
      t +
      0.254829592) *
      t *
      Math.exp(-x * x);
  return x >= 0 ? y : -y;
}

/** Standard normal CDF Φ(x). */
export function standardNormalCdf(x: number): number {
  return 0.5 * (1 + erf(x / Math.SQRT2));
}

/** BVC buy-volume fraction for a bucket, given its price change and σ. */
export function bvcBuyFraction(priceChange: number, sigma: number): number {
  if (!(sigma > 0)) return 0.5;
  return standardNormalCdf(priceChange / sigma);
}

/** Aggregate a trade stream into equal-volume buckets (splitting trades). */
export function bucketByVolume(
  trades: readonly { price: number; size: number }[],
  bucketSize: number,
): VolumeBucket[] {
  if (bucketSize <= 0) throw new Error("bucketSize must be positive");
  const buckets: VolumeBucket[] = [];
  let filled = 0;
  let prevClose = trades.length ? trades[0].price : 0;

  for (const t of trades) {
    let remaining = t.size;
    while (filled + remaining >= bucketSize) {
      remaining -= bucketSize - filled;
      buckets.push({ priceChange: t.price - prevClose, volume: bucketSize });
      prevClose = t.price;
      filled = 0;
    }
    filled += remaining;
  }
  return buckets;
}

function populationStd(xs: number[]): number {
  if (xs.length === 0) return 0;
  const mean = xs.reduce((a, b) => a + b, 0) / xs.length;
  const v = xs.reduce((a, b) => a + (b - mean) ** 2, 0) / xs.length;
  return Math.sqrt(v);
}

/**
 * VPIN over the last `window` buckets (default: all of them).
 * σ for BVC defaults to the std dev of the window's price changes.
 * Returns a value in [0, 1].
 */
export function vpin(
  buckets: VolumeBucket[],
  opts: { window?: number; sigma?: number } = {},
): number {
  if (buckets.length === 0) return 0;
  const window = opts.window ?? buckets.length;
  const slice = buckets.slice(-window);
  const sigma = opts.sigma ?? populationStd(slice.map((b) => b.priceChange));

  let imbalance = 0;
  let volume = 0;
  for (const b of slice) {
    const buyFrac = bvcBuyFraction(b.priceChange, sigma);
    imbalance += b.volume * Math.abs(2 * buyFrac - 1);
    volume += b.volume;
  }
  return volume === 0 ? 0 : imbalance / volume;
}

File: python/src/orderflow_metrics/vpin.py
"""VPIN — Volume-Synchronized Probability of Informed Trading.

Easley, Lopez de Prado & O'Hara (2012), "Flow Toxicity and Liquidity in a
High-Frequency World". Trades are grouped into equal-volume buckets; each bucket
is split into buy/sell volume by Bulk Volume Classification (BVC) using the
standardized close-to-close price change; VPIN is the average absolute order
imbalance across a rolling window of buckets::

    V_buy  = V * Phi(dP / sigma)
    V_sell = V * (1 - Phi(dP / sigma))
    VPIN   = sum |V_buy - V_sell| / sum V     over the window
"""
from __future__ import annotations

import math
from dataclasses import dataclass
from typing import List, Optional, Sequence

from .types import Trade


@dataclass(frozen=True)
class VolumeBucket:
    """An equal-volume bucket: close-to-close price change and its volume."""

    price_change: float
    volume: float


def _erf(x: float) -> float:
    # Abramowitz & Stegun 7.1.26 — max error ~ 1.5e-7.
    t = 1 / (1 + 0.3275911 * abs(x))
    y = 1 - (
        (
            (((1.061405429 * t - 1.453152027) * t + 1.421413741) * t - 0.284496736)
            * t
            + 0.254829592
        )
        * t
        * math.exp(-x * x)
    )
    return y if x >= 0 else -y


def standard_normal_cdf(x: float) -> float:
    """Standard normal CDF Phi(x)."""
    return 0.5 * (1 + _erf(x / math.sqrt(2)))


def bvc_buy_fraction(price_change: float, sigma: float) -> float:
    """BVC buy-volume fraction for a bucket, given its price change and sigma."""
    if not (sigma > 0):
        return 0.5
    return standard_normal_cdf(price_change / sigma)


def bucket_by_volume(trades: Sequence[Trade], bucket_size: float) -> List[VolumeBucket]:
    """Aggregate a trade stream into equal-volume buckets (splitting trades)."""
    if bucket_size <= 0:
        raise ValueError("bucket_size must be positive")
    buckets: List[VolumeBucket] = []
    filled = 0.0
    prev_close = trades[0].price if trades else 0.0

    for t in trades:
        remaining = t.size
        while filled + remaining >= bucket_size:
            remaining -= bucket_size - filled
            buckets.append(
                VolumeBucket(price_change=t.price - prev_close, volume=bucket_size)
            )
            prev_close = t.price
            filled = 0.0
        filled += remaining
    return buckets


def _population_std(xs: Sequence[float]) -> float:
    if not xs:
        return 0.0
    mean = sum(xs) / len(xs)
    v = sum((b - mean) ** 2 for b in xs) / len(xs)
    return math.sqrt(v)


def vpin(
    buckets: Sequence[VolumeBucket],
    window: Optional[int] = None,
    sigma: Optional[float] = None,
) -> float:
    """VPIN over the last ``window`` buckets (default: all of them).

    ``sigma`` for BVC defaults to the std dev of the window's price changes.
    Returns a value in [0, 1].
    """
    if not buckets:
        return 0.0
    w = window if window is not None else len(buckets)
    sliced = list(buckets)[-w:]
    s = sigma if sigma is not None else _population_std([b.price_change for b in sliced])

    imbalance = 0.0
    volume = 0.0
    for b in sliced:
        buy_frac = bvc_buy_fraction(b.price_change, s)
        imbalance += b.volume * abs(2 * buy_frac - 1)
        volume += b.volume
    return 0.0 if volume == 0 else imbalance / volume
